modules/supervisord.nix

```
@@ -0,0 +1,186 @@
+      };
+      autoRestart = lib.mkOption {
+        description = lib.mdDoc ''
+          Whether to automatically restart the process if it exits.
+
+          If `unexpected`, the process will be restarted if it exits
+          with an exit code not listed in the programs's `exitcodes`
+          configuration.
+        '';
+        type = types.either types.bool (types.enum [ "false" "true" "unexpected" ]);
+        default = "unexpected";
+      };
+      environment = lib.mkOption {
+        description = lib.mdDoc ''
+          Environment variables passed to the service's process.
+        '';
+        type = types.attrsOf types.str;
+        default = {};
+      };
+      extraConfig = lib.mkOption {
+        description = lib.mdDoc ''
+          Extra structured configurations to add to the [program:x] section.
+        '';
+        type = types.attrsOf types.str;
+        default = {};
+      };
+    };
+    config = {
+      command = lib.mkIf (config.script != "")
+        (toString (pkgs.writeShellScript "${name}-script.sh" config.script));
+
+      environment.PATH = lib.mkDefault (lib.makeBinPath config.path);
+    };
+  });
+
+  renderAtom = val:
+    if builtins.isBool val then if val then "true" else "false"
+    else toString val;
+
+  renderProgram = program: let
+    section = {
+      inherit (program) command;
+      autorestart = program.autoRestart;
+      environment = let
+        # FIXME: Make more robust
+        escape = builtins.replaceStrings [ "%" ] [ "%%" ];
+        envs = lib.mapAttrsToList (k: v: "${k}=\"${escape v}\"") program.environment;
+      in builtins.concatStringsSep "," envs;
+    } // program.extraConfig;
+  in lib.mapAttrs (_: v: renderAtom v) section;
+
+  numPrograms = builtins.length (builtins.attrNames cfg.programs);
+  enabledPrograms = lib.filterAttrs (_: program: program.enable) cfg.programs;
+
+  structuredConfig = {
+    supervisord = {
+      logfile = cfg.logPath;
+      pidfile = cfg.pidPath;
+    };
+    supervisorctl = {
+      serverurl = "unix://${cfg.socketPath}";
+    };
+    unix_http_server = {
+      file = cfg.socketPath;
+    };
+    "rpcinterface:supervisor" = {
+      "supervisor.rpcinterface_factory" = "supervisor.rpcinterface:make_main_rpcinterface";
+    };
+  } // (lib.mapAttrs' (k: v: {
+    name = "program:${k}";
+    value = renderProgram v;
+  }) enabledPrograms);
+
+  configFile = format.generate "supervisord.conf" structuredConfig;
+
+  # Only expose the "supervisorctl" executable
+  supervisorctl = pkgs.runCommand "supervisorctl" {} ''
+    mkdir -p $out/bin
+    ln -s ${cfg.package}/bin/supervisorctl $out/bin/supervisorctl
+  '';
+in {
+  options = {
+    supervisord = {
+      enable = lib.mkOption {
+        description = lib.mdDoc ''
+          Whether to enable the supervisord process control system.
+
+          This allows you to define long-running services in Nix-on-Droid.
+        '';
+        type = types.bool;
+        default = numPrograms != 0;
```

Wouldn't it make more sense to check for enabled programs only?
```suggestion
        default = enabledPrograms != 0;
```